x/shield/keeper/purchase.go

```diff
@@ -137,94 +144,141 @@ func (k Keeper) IterateAllPurchases(ctx sdk.Context, callback func(purchase type
 func (k Keeper) RemoveExpiredPurchases(ctx sdk.Context) {
 	store := ctx.KVStore(k.storeKey)
 	iterator := k.PurchaseQueueIterator(ctx, ctx.BlockTime())
-
+	bondDenom := k.sk.BondDenom(ctx)
 	defer iterator.Close()
 	for ; iterator.Valid(); iterator.Next() {
-		var timeslice []types.PurchaseTxHash
+		var timeslice []types.PPPair
 		k.cdc.MustUnmarshalBinaryLengthPrefixed(iterator.Value(), &timeslice)
-		for _, entry := range timeslice {
-			purchase, _ := k.GetPurchase(ctx, entry)
-			pool, err := k.GetPool(ctx, purchase.PoolID)
-			if err == nil {
-				pool.Available = pool.Available.Add(purchase.Shield.AmountOf(k.sk.BondDenom(ctx)))
-				pool.Shield = pool.Shield.Sub(purchase.Shield)
-				k.SetPool(ctx, pool)
+		for _, ppPair := range timeslice {
+			purchaseList, _ := k.GetPurchaseList(ctx, ppPair.PoolID, ppPair.Purchaser)
+
+			for i := 0; i < len(purchaseList.Entries); i++ {
+				entry := purchaseList.Entries[i]
+				if entry.ExpirationTime.Before(ctx.BlockTime()) {
+					purchaseList.Entries = append(purchaseList.Entries[:i], purchaseList.Entries[i+1:]...)
```

Comment:
By the way, is it ok to remove entries while iterating?

Reply:
And since order does not matter, to optimize performance what about doing:
```
purchaseList.Entries[i] = purchaseList.Entries[length-1]
return purchaseList.Entries[:length-1]
```
where `length = len(purchaseList.Entries)`

Reply:
Should be fine https://play.golang.org/p/zjSjRuHIB4u

Reply:
Oh maybe not https://play.golang.org/p/ye6jHg1FMeT

Reply:
We can do it from `len(purchaseList.Entries)-1` to `0`.

Reply:
or decrement `i` after deleting an entry

Reply:
`m = append(m[:i], m[i+1:]...)` is fine though it does do a copy on every element to be removed. Personally I'd go with something more idiomatic. Such as simply appending the to-be-saved elements into a new slice.

Reply:
yeah as long as we keep the index on track. Forgot to put `i--` in there. Seemed to work fine without it though 🤔